src/replication.c

```diff
@@ -351,50 +375,40 @@ void clear_replica_cmd(spec_t *spec, replica_t *replica, rcmd_t *rep_cmd) {
 	}
 }
 
-
-int
-remove_replica_from_list(spec_t *spec, int iofd) {
-	replica_t *replica;
-	int ios_aborted = 0, rc;
+void
+remove_replica_from_list(spec_t *spec, replica_t *replica) {
+	int ios_aborted = 0, rep_id;
 	rcmd_t *rep_cmd = NULL;
-	struct epoll_event event;
 	MTX_LOCK(&spec->rq_mtx);
-	TAILQ_FOREACH(replica, &spec->rq, r_next) {
-		if(iofd == replica->iofd) {
-			REPLICA_LOG("REMOVE REPLICA FROM LIST\n");
-			MTX_LOCK(&replica->r_mtx);
-			replica->state = REPLICA_ERRORED;
-			//Empty waitq of replica
-			while((rep_cmd = TAILQ_FIRST(&replica->waitq))) {
-				clear_replica_cmd(spec, replica, rep_cmd);
-				TAILQ_REMOVE(&replica->waitq, rep_cmd, rwait_cmd_next);
-				ios_aborted++;
-			}
-			//Empty blockedq of replica
-			while((rep_cmd = TAILQ_FIRST(&replica->blockedq))) {
-				clear_replica_cmd(spec, replica, rep_cmd);
-				TAILQ_REMOVE(&replica->blockedq, rep_cmd, rblocked_cmd_next);
-				ios_aborted++;
-			}
-			replica->removed = true;
-			pthread_cond_signal(&replica->r_cond);
-			MTX_UNLOCK(&replica->r_mtx);
-			spec->replica_count--;
-			//TODO Update spec->degraded_rcount || spec->healthy_rcount over here
-			//based on state of  replica when it gets disconnected
-			update_volstate(spec);
-			TAILQ_REMOVE(&spec->rq, replica, r_next);
-			rc = epoll_ctl(spec->receiver_epfd, EPOLL_CTL_DEL, replica->iofd, &event);
-			if (rc == -1) {
-				return -1;
-			}
-			close(replica->iofd);
-			close(replica->sender_epfd);
-		}
+	MTX_LOCK(&replica->r_mtx);
+	REPLICA_LOG("REMOVE REPLICA FROM LIST\n");
+	if(replica->state == REPLICA_DEGRADED)
+		spec->degraded_rcount--;
+	if(replica->state == REPLICA_HEALTHY)
+		spec->healthy_rcount--;
+	update_volstate(spec);
```

Comment:
do tailq_remove of replica from rq_list, and unlock the spec->rq_mtx here itself, but, can be before taking r_mtx lock

Reply:
basically, follow the reverse order of adding replica entry (i.e., on update_replica_entry).. so, it should be:
- remove from list, update vol state
- stop the threads
- close the sockets

Reply:
lets have protocol that a replica will be in spec's list only if it is in DEGRADED or HEALHTY state..
so, I mean that - replica need to be removed from list before changing its state to ERRORED